pkg/webhook/defaults/ingress.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2020-2021 Clastix Labs
+// SPDX-License-Identifier: Apache-2.0
+
+package defaults
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	capsulev1beta2 "github.com/clastix/capsule/api/v1beta2"
+	capsuleingress "github.com/clastix/capsule/pkg/webhook/ingress"
+	"github.com/clastix/capsule/pkg/webhook/utils"
+)
+
+func mutateIngressDefaults(ctx context.Context, req admission.Request, c client.Client, decoder *admission.Decoder, recorder record.EventRecorder) *admission.Response {
+	ingress, err := capsuleingress.FromRequest(req, decoder)
+	if err != nil {
+		return utils.ErroredResponse(err)
+	}
+
+	var tnt *capsulev1beta2.Tenant
+
+	tnt, err = capsuleingress.TenantFromIngress(ctx, c, ingress)
+	if err != nil {
+		return utils.ErroredResponse(err)
+	}
+
+	if tnt == nil {
+		return nil
+	}
+
+	// Validate Default Ingress
+	allowed := tnt.Spec.IngressOptions.AllowedClasses
+
+	if allowed == nil || allowed.Default == "" {
+		return nil
+	}
+
+	mutate := false
+
+	var ingressClass client.Object
+
+	ingressClassName := ingress.IngressClass()
+
+	if ingressClassName != nil {
+		if ingressClassName != &allowed.Default {
+			ingressClass, err = utils.GetIngressClassByName(ctx, c, ingressClassName)
+			if err != nil && !k8serrors.IsNotFound(err) {
+				response := admission.Denied(NewIngressClassError(*ingressClassName, err).Error())
+
+				return &response
+			}
+		}
+	} else {
+		mutate = true
+	}
+
+	// If given class is cluster default
+	if ingressClass != nil && utils.IsDefaultIngressClass(ingressClass) {
+		// If cluster default is not allowed in tenant
+		if !allowed.MatchSelectByName(ingressClass) {
+			mutate = true
+		}
+	}
+
+	if mutate {
+		ingress.SetIngressClass(allowed.Default)
+		recorder.Eventf(tnt, corev1.EventTypeNormal, "TenantDefault", "Assigned Tenant default Ingress Class %s to %s/%s", allowed.Default, ingress.Name(), ingress.Namespace())
```

Comment:
We should record only if the handler is returning with no error, still a chance for an error at line 80.